### PR TITLE
Make nil-tolerable #try_unwrap

### PR DESCRIPTION
**Pull request by Kukunin**

### CODE REVIEW

**Comment by solnic** [comment]
I'd accept something like `safe_unwrap` and `safe_wrap` with this kind of logic, we shouldn't have transformations that go silently through potentially broken data

**Comment by Kukunin** [comment]
What's about `try_unwrap` ?

**Comment by solnic** [comment]
@Kukunin hmm yeah `try_unwrap` sounds better

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 ## Added
 
 * prefix option to HashTransformations#unwrap and HashTransformations#unwrap! (AMHOL)
+* `try_unwrap` transformation, which works with `nil` or absent values (Kukunin)
```
Comment (AMHOL) [inline CHANGELOG.md]: This was added to the previous release section

File: lib/transproc/hash.rb [hunk 2]
```diff
@@ -333,6 +333,23 @@ def self.unwrap(hash, root, keys = nil, prefix: false)
...
 
+    # Same as `:unwrap` but handles nil values
+    #
+    # If value for a specified key is nil, or there is no such key in a hash
+    # `:try_unwrap` returns hash without the specified key and with no added element
+    #
+    # @see HashTransformations.unwrap
+    #
+    # @api public
+    def self.try_unwrap(hash, root, keys = nil, prefix: false)
```
Comment (AMHOL) [inline lib/transproc/hash.rb]: Could we also add a `try_unwrap!` mutating version? We have it as standard for all other hash transformations, I know @solnic mentioned we were looking at getting rid of them but I think it's best to be consistent until we do

Comment (Kukunin) [inline lib/transproc/hash.rb]: btw, what's the application of mutating versions? `transproc` feeds output of method to the next method, so mutating isn't even needed

**Comment by Kukunin** [comment]
Improved

File: lib/transproc/hash.rb (hunk 2 above)
Comment (AMHOL) [inline lib/transproc/hash.rb]: Yeah I don't think anyone actually uses it, but it was just added in the initial version so everyone stuck with the convention, guess we didn't consider YAGNI :laughing:

**Comment by Kukunin** [comment]
I've just understood, that I need something slightly different than `unwrap`, and `try_unwrap` isn't suitable as well. 

For SQL mapping, I need to get some attributes from nested objects, and remove original object.
In most cases, I need to get `#id` attribute, but I may need to get 2-3 attributes (to extract value object to store inline).
In all cases, there shouldn't be original object in hash, because we don't have such column in database. If there is no nested object in hash or it's `nil`, it should handle and remove it anyway. 

@solnic did you get it? with your better knowledge of terminology, do you have any idea about how to call such operation?

File: lib/transproc/hash.rb (hunk 2 above)
Comment (solnic) [inline lib/transproc/hash.rb]: rom used use mutating functions for perf reasons but then I saw some really odd bugs so I remove their usage from rom (who would have thought? ;)).

They should be removed before 1.0.0

**Comment by solnic** [comment]
@Kukunin so it seems like you're talking about a composition of wrap + remove_keys

**Comment by Kukunin** [comment]
so it's literally `t(:is, Hash, t(:unwrap, :address, [:street, :city, :state])) + (t(:reject_keys, [:address]))`

but do we have any plans to extend built-in transformations with more complex? 
Do we have any plans to figure out some naming conventions etc? 

What's your opinion?

**Comment by Kukunin** [comment]
It'd cool to have some DSL in registry to compose more complex transformations:

```ruby
module T
  extend Transproc::Registry

  compose :safe_unwrap do
    t(:is, Hash, t(:unwrap, :address, [:street, :city, :state])) >> (t(:reject_keys, [:address]))
  end
end
```

because currently it looks like 

```ruby
module T
  extend Transproc::Registry

  def self.safe_unwrap(value)
    t(:is, Hash, t(:unwrap, :address, [:street, :city, :state]))
      .compose(t(:reject_keys, [:address]))
      .call(value)
  end
end
```

**Pull request closed without merging**